lang/src/org/partiql/lang/syntax/SqlParser.kt

```diff
@@ -2246,6 +2267,135 @@ class SqlParser(private val ion: IonSystem) : Parser {
+     */
+    private fun List<Token>.parseOptionalPrecision(): ParseNode =
+        // If the optional precision is present
+        if (head?.type == LEFT_PAREN) {
+            var rem = tail
+            // Expected precision token to be unsigned integer
+            if (rem.head == null || rem.head!!.type != LITERAL || !rem.head!!.value!!.isUnsignedInteger) {
+                rem.head.err("Expected integer value for precision", PARSE_INVALID_PRECISION_FOR_TIME)
+            }
+            val precision = rem.head
+            rem = rem.tail
+            if (rem.head?.type != RIGHT_PAREN) {
+                rem.head.errExpectedTokenType(RIGHT_PAREN)
+            }
+            ParseNode(PRECISION, precision, listOf(), rem.tail)
+        }
+        else {
+            ParseNode(PRECISION, null, listOf(), this)
+        }
+
+    /**
+     * Checks for the optional time zone indicator which is specified with "WITH TIME ZONE"
+     */
+    private fun List<Token>.checkForOptionalTimeZone(): Pair<List<Token>, Boolean> {
+        // If the keyword is specified for time zone, it must be a series of keywords - "with time zone"
+        if (head?.type == KEYWORD) {
+            val rem =
+                tailExpectedKeyword("with").
+                tailExpectedKeyword("time").
+                tailExpectedKeyword("zone")
+            return Pair(rem, true)
+        }
+        return Pair(this, false)
+    }
+
+    /**
+     * Parses a time string and verifies that the time string is a string ond is specified in the valid ISO format.
+     * Allows for optional precision and time zone to be specified with the time.
+     * The different valid usages are as follows:
+     *
+     *         1. TIME 'HH:MM:SS[.ddd....][+|-HH:MM]'
+     *
+     *         2. TIME (<p>) 'HH:MM:SS[.ddd....][+|-HH:MM]'
+     *
+     *         3. TIME WITH TIME ZONE 'HH:MM:SS[.ddd....][+|-HH:MM]'
+     *
+     *         4. TIME (<p>) WITH TIME ZONE 'HH:MM:SS[.ddd....][+|-HH:MM]'
+     *
+     * where p is the precision for the second's value in the time.
+     * The valid range for the time zone offset is [-18:00 to +18:00]
+     * If the time zone offset is not specified when used with "TIME WITH TIME ZONE", the system default time zone is picked.
+     * This behaviour is consistent with most other database languages like SQL92, PostgreSQL etc.
+     * It also makes more sense as the user using the keywords "TIME WITH TIME ZONE .." would expect the
+     * local time zone offset to be used by default.
+     */
+    private fun List<Token>.parseTime(): ParseNode {
+
+        // These are used to validate the generic format of the time string.
+        // The more involved logic such as validating the time is done by LocalTime.parse or OffsetTime.parse
+        val timeWithoutTimeZoneRegex = Regex("\\d\\d:\\d\\d:\\d\\d(\\.\\d*)?")
+        val genericTimeRegex = Regex("\\d\\d:\\d\\d:\\d\\d(\\.\\d*)?([+|-]\\d\\d:\\d\\d)?")
```

Comment:
I think you can move these to private top-level `private` variables since `Regex` implementations are thread-safe.

Reply:
Done.